synapse/cryotank.py

```diff
@@ -571,3 +588,610 @@ def init(self, name, conf=None, timeout=None):
+        self._workq = s_queue.Queue()
+        # A dict of propname -> version, type, datapath dict
+        self._meta = _IndexMeta(self._dbenv)
+        self._next_offset = self._meta.lowestProgress()
+        self._chunk_sz = 1000  # < How many records to read at a time
+        self._remove_chunk_sz = 1000  # < How many index entries to remove at a time
+        ebus.on('cryotank:puts', self._onData)
+
+        self._worker.start()
+
+        def _onfini():
+            self._workq.done()
+            self._worker.join(self.MAX_WAIT_S)
+            self._dbenv.close()
+
+        ebus.onfini(_onfini)
+
+    def _onData(self, unused):
+        '''
+        Wake up the worker if he already doesn't have a reason to be awake
+        '''
+        if 0 == len(self._workq):
+            self._workq.put((None, lambda: None, None, None))
+
+    def _removeSome(self):
+        '''
+        Make some progress on removing deleted indices
+        '''
+        left = self._remove_chunk_sz
+        for iid in self._meta.deleting:
+            if not left:
+                break
+            iid_enc = _iid_en(iid)
+            with self._dbenv.begin(db=self._idxtbl, buffers=True, write=True) as txn, txn.cursor() as curs:
+                if curs.set_range(iid_enc):
+                    for k, offset_enc in curs.iternext():
+                        if k[:len(iid_enc)] != iid_enc:
+                            break
+                        if not curs.delete():
+                            raise s_exc.CorruptDatabase('delete failure')
+
+                        txn.delete(offset_enc, iid_enc, db=self._normtbl)
+                        left -= 1
+                        if not left:
+                            break
+                if not left:
+                    break
+
+            self._meta.markDeleteComplete(iid)
+
+    def _normalize_records(self, raw_records):
+        '''
+        Yields stream of normalized fields
+
+        Args:
+            raw_records(Iterable[Tuple[int, Dict[int, str]]])  generator of tuples of offset/decoded raw cryotank
+            record
+        Returns:
+            Iterable[Tuple[int, int, Union[str, int]]]: generator of tuples of offset, index ID, normalized property
+            value
+        '''
+        for offset, record in raw_records:
+            self._next_offset = offset + 1
+            dp = s_datapath.initelem(s_msgpack.un(record))
+
+            for iid, idx in ((k, v) for k, v in self._meta.indices.items() if not self._meta.asleep[k]):
+                if self._meta.progresses[iid]['nextoffset'] > offset:
+                    continue
+                try:
+                    self._meta.progresses[iid]['nextoffset'] = offset + 1
+                    for datapath in idx.datapath:
+                        field = dp.valu(datapath)
+                        if field is None:
+                            continue
+                        # TODO : what to do with subprops?
```

Comment:
Does this still need to be handled?

Reply:
It won't break anything. I have a top-level TODO about it.